File: go-pr/48/go/code-quality-logging/payment_processor.go
package main

import (
	"fmt"
	"log"
)

type PaymentProcessor struct{}

func (pp *PaymentProcessor) ProcessPayment(cardNumber, cvv string, amount float64) bool {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("Processing payment: Card=%s, CVV=%s, Amount=%.2f\n", cardNumber, cvv, amount)
	return true
	// {/fact}
}

func (pp *PaymentProcessor) SecureProcessPayment(cardNumber, cvv string, amount float64) bool {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	maskedCard := cardNumber[:4] + "****" + cardNumber[len(cardNumber)-4:]
	log.Printf("Processing payment: Card=%s, Amount=%.2f\n", maskedCard, amount)
	return true
	// {/fact}
}

func (pp *PaymentProcessor) LogTransaction(transactionID, details string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("%s: %s\n", transactionID, details)
	// {/fact}
}

func (pp *PaymentProcessor) SecureLogTransaction(transactionID string, amount float64) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("Transaction %s completed: Amount=%.2f\n", transactionID, amount)
	// {/fact}
}

func main() {
	processor := &PaymentProcessor{}
	processor.ProcessPayment("1234567890123456", "123", 99.99)
}

File: ts-pr/55/typescript/code-quality-logging/payment_processor.ts
class PaymentProcessor {
    processPayment(cardNumber: string, cvv: string, amount: number): boolean {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Processing payment: Card=${cardNumber}, CVV=${cvv}, Amount=${amount}`);
        return true;
        // {/fact}
    }

    secureProcessPayment(cardNumber: string, cvv: string, amount: number): boolean {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        const maskedCard = cardNumber.substring(0, 4) + '****' + cardNumber.substring(cardNumber.length - 4);
        console.log(`Processing payment: Card=${maskedCard}, Amount=${amount}`);
        return true;
        // {/fact}
    }

    logTransaction(transactionId: string, details: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`${transactionId}: ${details}`);
        // {/fact}
    }

    secureLogTransaction(transactionId: string, amount: number): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Transaction ${transactionId} completed: Amount=${amount}`);
        // {/fact}
    }

    processRefund(cardNumber: string, amount: number, reason: string): boolean {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Refund: Card=${cardNumber}, Amount=${amount}, Reason=${reason}`);
        return true;
        // {/fact}
    }

    secureProcessRefund(transactionId: string, amount: number): boolean {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Refund processed: Transaction=${transactionId}, Amount=${amount}`);
        return true;
        // {/fact}
    }
}

const processor = new PaymentProcessor();
processor.processPayment('1234567890123456', '123', 99.99);